kafkakewl-metrics/src/main/scala/com/mwam/kafkakewl/metrics/services/KafkaConsumerGroupMetricsCalc.scala

```diff
@@ -0,0 +1,263 @@
+          .getOrElse(ZIO.unit)
+      }
+
+      // Create new workers for new ConsumerGroupTopicPartition
+      addedWorkers <- ZIO.foreachPar(addedKeys) { consumerGroupTopicPartition =>
+        ZIO
+          .succeed(consumerGroupTopicPartition)
+          .zip(createNewWorkerFiber(consumerGroupTopicPartition, latestTopicPartitionInfos, newLatestConsumerGroupOffsets))
+      }
+
+      workersWithAdded = workers ++ addedWorkers
+
+      // Update partitionConsumers to reflect new and removed consumer groups
+      newPartitionConsumers = ConsumerGroupMetricsManager.updateMapOfSetsWithAddedRemoved(
+        partitionConsumers,
+        addedOrUpdated.map((cgtp, _) => (cgtp.topicPartition, cgtp)),
+        removed.map(cgtp => (cgtp.topicPartition, cgtp)).toMap
+      )
+
+      // Remove workers that have had nulls pushed to them, interrupting their fibres
+      _ <- ZIO.foreachParDiscard(removedGroupTopicPartition) { groupTopicPartition =>
+        workersWithAdded.get(groupTopicPartition).map(worker => worker.fibreHandle.interrupt).getOrElse(ZIO.unit) *> workerOutputQueue.offer(
+          KafkaConsumerGroupMetricChanges(Map.empty, Set(groupTopicPartition))
+        )
+      }
+      workersWithRemoved = workersWithAdded.removedAll(removedGroupTopicPartition)
+
+    } yield ConsumerGroupMetricsManager(
+      workerOutputQueue,
+      workersWithRemoved,
+      newPartitionConsumers,
+      latestTopicPartitionInfos,
+      newLatestConsumerGroupOffsets
+    )
+  }
+
+  private def createNewWorkerFiber(
+      consumerGroupTopicPartition: ConsumerGroupTopicPartition,
+      latestTopicPartitionInfos: Map[KafkaTopicPartition, KafkaTopicPartitionInfo],
+      latestConsumerGroupOffsets: Map[ConsumerGroupTopicPartition, KafkaConsumerGroupOffset]
+  ): UIO[ConsumerGroupTopicPartitionMetricsWorker] = {
+    for {
+      partitionInfoUpdate <- Queue.unbounded[KafkaTopicPartitionInfo]
+      _ <- ZIO.foreachDiscard(latestTopicPartitionInfos.get(consumerGroupTopicPartition.topicPartition))(topicPartitionInfo =>
+        partitionInfoUpdate.offer(topicPartitionInfo)
+      )
+      consumerGroupUpdate <- Queue.unbounded[KafkaConsumerGroupOffset]
+      _ <- ZIO.foreachDiscard(latestConsumerGroupOffsets.get(consumerGroupTopicPartition))(consumerGroupOffset =>
+        consumerGroupUpdate.offer(consumerGroupOffset)
+      )
+
+      workerFiber <- ZStream
+        .fromQueueWithShutdown(partitionInfoUpdate)
+        .mergeEither(ZStream.fromQueueWithShutdown(consumerGroupUpdate))
+        .runFoldZIO((Option.empty[KafkaTopicPartitionInfo], Option.empty[KafkaConsumerGroupOffset])) { (state, update) =>
+          // TODO: The window metric functions
+          val (newState) = update match {
+            case Left(kafkaTopicPartitionInfo)   => (Some(kafkaTopicPartitionInfo), state._2)
+            case Right(kafkaConsumerGroupOffset) => (state._1, Some(kafkaConsumerGroupOffset))
+          }
+          workerOutputQueue
```

Comment:
nitpick, may deconstructing newState into its parts `val (someGoodName, otherGoodName) = newState` makes it more readable below where you refer to `_1` and `_2`

Reply:
Good point, this is done